### PR TITLE
16502 refactor rpcprotocol

### PR DESCRIPTION
**Pull request by tsmaeder**
### What does this PR do?
This PR adapts che-theia to the upstream PR https://github.com/eclipse-theia/theia/pull/8972. As such, I'll leave it in draft PR staus until upstream is merged. It is still open for review.

### What issues does this PR fix or reference?
This is a refactoring and as such not related to a PR. See upstream PR for justification.

### How to test this PR?
Use local and remote plugins in a workspace based on the results of this PR build. (self-hosting or via the image built for the happy-path tests.

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [x] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### Happy Path Channel

HAPPY_PATH_CHANNEL=next

### CODE REVIEW

**Comment by tsmaeder** [comment]
Since I have two thumbs up on the upstream PR, I'm removing the "draft" status, Reviewers, please review!

File: extensions/eclipse-che-theia-plugin-remote/src/node/hosted-plugin-remote.ts [hunk 1]
```diff
@@ -50,6 +50,7 @@ export class HostedPluginRemote {
 
   @postConstruct()
   protected postConstruct(): void {
+    console.log('hosted plugin remote postconstruct');
```
Comment (azatsarynnyy) [inline extensions/eclipse-che-theia-plugin-remote/src/node/hosted-plugin-remote.ts:53]: Do we need to keep this log or it's just a leftover?

**Review by azatsarynnyy — COMMENTED** [review]
The code looks good to me, as it's just an adaptation to the upstream refactoring.
Let us know once a Che-Theia image is ready for testing the PR.

**Comment by tsmaeder** [comment]
> Let us know once a Che-Theia image is ready for testing the PR.

I honestly don't know how to create one: the Github action PR checks don't publish images and the "happy path" CI job doesn't have the ability to work with Theia branches. 🤷  I guess self-hosting is the process to test in this case.

**Comment by benoitf** [comment]
@tsmaeder you could change in build.include file the branch used

**Comment by tsmaeder** [comment]
Hmh...does the plugin-host executable get rebuilt when we rebuild the che-theia image? Not sure anymore. Because the protocol between back-end and plugin host changed. @benoitf what would be the right way to test that?

**Comment by benoitf** [comment]
@tsmaeder yes it builds the endpoint image as well

Right way: tune build.include to include remote upstream theia branch/fork instead of eclipse-theia/theia + master so Happy path will deploy both custom theia and custom theia endpoint (so it will validate if communication across containers/VS Code extensions works fine)

**Comment by tsmaeder** [comment]
[crw-ci-test]

**Comment by tsmaeder** [comment]
[crw-ci-test --rebuild]

**Comment by tsmaeder** [comment]
@benoitf @azatsarynnyy we have an image being built now. Could you test, please?

**Comment by benoitf** [comment]
let's merge it after 7.28.0 is tagged (ie: for 7.29.0 so it'll be a new CRW/Che lifecycle)

**Review by azatsarynnyy — APPROVED** [review]

**Comment by ericwill** [comment]
@tsmaeder what's the status of this PR?

**Comment by tsmaeder** [comment]
When I get a moment, I need to rebase the upstream PR before I can merge both.

**Comment by benoitf** [comment]
now, we're at 2/3 of the sprint (my previous comment was 3 weeks ago) so it's not becoming a good time to merge refactoring upstream/downstream stuff

File: build.include [hunk 2]
```diff
@@ -11,9 +11,9 @@ set -e
 set -u
 
 IMAGE_TAG="next"
-THEIA_GITHUB_REPO="eclipse-theia/theia"
+THEIA_GITHUB_REPO="tsmaeder/theia"
```
Comment (vitaliy-guliy) [inline build.include:14]: I suppose this change and `THEIA_BRANCH` will be reverted before merge.

**Comment by tsmaeder** [comment]
[crw-ci-test --rebuild]

**Pull request merged**